src/main/java/jenkins/scm/api/SCMEvent.java

```diff
@@ -247,14 +309,83 @@ public int hashCode() {
     @Override
     public String toString() {
         return String.format(
-                "SCMEvent{type=%s, timestamp=%tc, payload=%s}",
+                "SCMEvent{type=%s, timestamp=%tc, payload=%s, origin=%s}",
                 type,
                 timestamp,
-                payload
+                payload,
+                StringUtils.defaultIfBlank(origin, ORIGIN_UNKNOWN)
         );
     }
 
     /**
+     * Helper method to get the origin of an event from a {@link HttpServletRequest}. The current format is the
```

Comment:
If it helper and not private, why not put it into helper class? Scrolling class with helpers and nested classes makes eyes red. Though https://github.com/jenkinsci/ldap-plugin/blob/master/src/main/java/hudson/security/LDAPSecurityRealm.java is still record with 15 classes.